smart-contracts/contracts/cl-vault/src/contract.rs

```diff
@@ -177,13 +190,149 @@ pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractEr
+        .add_attribute("liquidity_amount", position.liquidity))
+}
+
+fn handle_migration_withdrawal_reply(
+    deps: DepsMut,
+    env: Env,
+    data: SubMsgResult,
+) -> ContractResult<Response> {
+    //remove old positions
+    POSITION.remove(deps.storage);
+
+    let migration_data = MIGRATION_DATA.load(deps.storage)?;
+    // Step 2: Verify pool assets if they match
+    let new_pool: Pool = PoolmanagerQuerier::new(&deps.querier)
+        .pool(migration_data.new_pool_id)?
+        .pool
+        .ok_or(ContractError::PoolNotFound {
+            pool_id: migration_data.new_pool_id,
+        })?
+        .try_into()
+        .unwrap();
+
+    let pool_config = POOL_CONFIG.load(deps.storage)?;
+
+    if (new_pool.token0 != pool_config.token0) || (new_pool.token1 != pool_config.token1) {
+        return Err(ContractError::PoolTokenMismatch {});
+    }
+
+    //calculate total assets after withdrawal
+    let msg: MsgWithdrawPositionResponse = data.try_into()?;
+
+    let mut amount0: Uint128 = msg.amount0.parse()?;
+    let mut amount1: Uint128 = msg.amount1.parse()?;
+
+    let unused_balances = get_unused_balances(deps.storage, &deps.querier, &env)?;
+
+    let unused_balance0 = unused_balances
+        .find_coin(pool_config.token0.clone())
+        .amount
+        .checked_sub(amount0)?;
+    let unused_balance1 = unused_balances
+        .find_coin(pool_config.token1.clone())
+        .amount
+        .checked_sub(amount1)?;
+
+    amount0 = amount0.checked_add(unused_balance0)?;
+    amount1 = amount1.checked_add(unused_balance1)?;
+
+    CURRENT_BALANCE.save(deps.storage, &(amount0, amount1))?;
+
+    let mut tokens_provided = vec![];
+    if !amount0.is_zero() {
+        tokens_provided.push(Coin {
+            denom: pool_config.token0,
+            amount: amount0,
+        })
+    }
+    if !amount1.is_zero() {
+        tokens_provided.push(Coin {
+            denom: pool_config.token1,
+            amount: amount1,
+        })
+    }
+
+    //save new pool info
+    POOL_CONFIG.save(
+        deps.storage,
+        &PoolConfig {
+            pool_id: new_pool.id,
+            token0: new_pool.token0,
+            token1: new_pool.token1,
+        },
+    )?;
+
+    // create position in new pool
+    let create_positon_msg = create_position(
+        deps,
+        &env,
+        migration_data.lower_tick,
+        migration_data.upper_tick,
+        tokens_provided,
+        Uint128::zero(),
+        Uint128::zero(),
+    )?;
+
+    Ok(Response::new()
+        .add_submessage(SubMsg::reply_on_success(
+            create_positon_msg,
+            Replies::CreateMigratedPosition.into(),
+        ))
+        .add_attribute("migration", "in_progress")
+        .add_attribute("position", "created"))
+}
+
+pub fn handle_create_migrated_position_reply(
+    deps: DepsMut,
+    _env: Env,
+    data: SubMsgResult,
+) -> Result<Response, ContractError> {
+    let response: MsgCreatePositionResponse = data.try_into()?;
+
```

Comment:
nitpick but it would be good to destroy migration data after this, wdyt?

Reply:
yes agreed, it's cleaner

Reply:
pushing